src/promql/src/functions/extrapolate_rate.rs

```diff
@@ -0,0 +1,564 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementations of `rate`, `increase` and `delta` functions in PromQL.
+
+use std::fmt::Display;
+use std::sync::Arc;
+
+use datafusion::arrow::array::{Float64Array, TimestampMillisecondArray};
+use datafusion::arrow::datatypes::TimeUnit;
+use datafusion::common::DataFusionError;
+use datafusion::logical_expr::{ScalarUDF, Signature, TypeSignature, Volatility};
+use datafusion::physical_plan::ColumnarValue;
+use datatypes::arrow::array::Array;
+use datatypes::arrow::datatypes::DataType;
+
+use crate::extension_plan::Millisecond;
+use crate::functions::extract_array;
+use crate::range_array::RangeArray;
+
+pub type Delta = ExtrapolatedRate<false, false>;
+pub type Rate = ExtrapolatedRate<true, true>;
+pub type Increase = ExtrapolatedRate<true, false>;
+
+/// Part of the `extrapolatedRate` in Promql,
+/// from https://github.com/prometheus/prometheus/blob/6bdecf377cea8e856509914f35234e948c4fcb80/promql/functions.go#L66
+#[derive(Debug)]
+pub struct ExtrapolatedRate<const IS_COUNTER: bool, const IS_RATE: bool> {
+    /// Range duration in millisecond
+    range_length: i64,
+}
+
+impl<const IS_COUNTER: bool, const IS_RATE: bool> ExtrapolatedRate<IS_COUNTER, IS_RATE> {
+    /// Constructor. Other public usage should use [`scalar_udf`] instead.
+    fn new(range_length: i64) -> Self {
+        Self { range_length }
+    }
+
+    fn input_type() -> Vec<DataType> {
+        vec![
+            // timestamp range vector
+            RangeArray::convert_data_type(DataType::Timestamp(TimeUnit::Millisecond, None)),
+            // value range vector
+            RangeArray::convert_data_type(DataType::Float64),
+            // timestamp vector
+            DataType::Timestamp(TimeUnit::Millisecond, None),
+        ]
+    }
+
+    fn return_type() -> DataType {
+        DataType::Float64
+    }
+
+    fn calc(&self, input: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+        assert_eq!(input.len(), 3);
+
+        // construct matrix from input
+        let ts_array = extract_array(&input[0])?;
+        let ts_range = RangeArray::try_new(ts_array.data().clone().into())?;
+        let value_array = extract_array(&input[1])?;
+        let value_range = RangeArray::try_new(value_array.data().clone().into())?;
+        let ts = extract_array(&input[2])?;
+        let ts = ts
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
```

Comment:
So we don't support time index type to be bigint?

Reply:
Yes. At this stage all types are hardcoded. Timestamp can only be `TimestampMillisecond` 😥